source/postprocess/visualization/heat_flux_map.cc

```diff
@@ -0,0 +1,176 @@
+                    for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                      in.composition[i][c] = composition_values[c][i];
+                  }
+
+                this->get_material_model().evaluate(in, out);
+
+                // Calculate the normal conductive heat flux given by the formula
+                //   j = - k * n . grad T
+
+                double normal_flux = 0;
+                for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+                  {
+                    const double thermal_conductivity
+                      = out.thermal_conductivities[q];
+
+                    normal_flux += -thermal_conductivity *
+                                   (temperature_gradients[q] * fe_face_values.normal_vector(q)) *
+                                   fe_face_values.JxW(q);
+                  }
+
+                // get the location for each gridpoint at the top boundary
+                fe_face_values.reinit(cell, top_face_idx);
+                const Point<dim> midpoint_at_surface = cell->face(top_face_idx)->center();
+
+                (*return_value.second)(cell_index) = normal_flux;
+
+              }
+
+        return return_value;
+      }
+
+      template <int dim>
+      void
+      HeatFluxMap<dim>::
+      declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+        }
+        prm.leave_subsection();
+      }
+
+      template <int dim>
+      void
+      HeatFluxMap<dim>::parse_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+        }
+        prm.leave_subsection();
+      }
+    }
+  }
+}
+
+
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(HeatFluxMap,
+                                                  "heat flux map",
+                                                  "A visualization output object that generates output for "
+                                                  "the heatflux across the top boundary. The heat flux is computed in "
+                                                  "outward direction, i.e., from the domain to the outside, "
+                                                  "using the formula $k \\nabla T \\cdot \\mathbf n$, where "
+                                                  "$k$ is the thermal conductivity as reported by the material "
+                                                  "model, $T$ is the temperature, and $\\mathbf n$ is the outward "
+                                                  "normal. Note that the quantity so computed does not include "
+                                                  "any energy transported across the boundary by material "
+                                                  "transport in cases where $\\mathbf u \\cdot \\mathbf n \\neq 0$.")
```

Comment:
maybe add a sentence that the *integrated* heatflux for each part of the boundary can be obtained by the "heat flux" postprocessor; maybe also add a sentence there that the *pointwise* heat flux can be obtained using the "heat flux map" visualization postprocessor